---
title: "single_cell_seurat_2"
author: "Hannah Pook"
date: '2022-05-24'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(Seurat)
library(tidyverse)
library(patchwork)
library(DT)
library(gprofiler2)

```

# Exercise 1

## Read in the RDS object we created and save from Seurat day 1

- use the `readRDS()` function to read in previously saved object

```{r}
# object already loaded so did not need to use readRDS this time
head(seurat_after_qc)
```

- Check the Default Assay

```{r}
DefaultAssay(seurat_after_qc)
```


## Apply SCTransfrom  normalisation

Use `SCTransform()` function

SCTransform vignette: https://satijalab.org/seurat/articles/sctransform_vignette.html 

`SCTransform()`  command replaces `NormalizeData()`, `ScaleData` and `FindVariableFeatures()` run for the RNA assay in day 1 Seurat

Should we remove any confounding variables like we did for the RNA assay for Day 1?

Do we want to use the same number of variable featuresn(n=1000) or more than what we used for `NormalizeData()` function.

> As SCTransform models the variable genes better than log normalisation we can use a higher number of variable genes compared to log normalisation (here we are using 3000 compared to 2000 with log normalisation)

> In vars.to.regress we do not need to add UMI as the function is already normalising for sequencing depth! (This is in contrast to what we did yesterday with log normalisation method)

> However, we do still want to regress out percent mito (like we did yesterday with log normalisation method)


```{r}
seurat_after_qc <- SCTransform(
  object = seurat_after_qc,
  assay = "RNA",
  new.assay.name = "SCT",
  do.correct.umi = TRUE,
  ncells = 5000,
  variable.features.n = 3000,
  vars.to.regress = "percent_mt"
)
```

Where is the new normalisation stored?
> Answer:
> The new normalisation is stored in a new assay called "SCT"
> When run function can see that SCT has been set to default assay 

Explore the `seurat_after_qc` objects meta.data and assays


```{r}
Assays(seurat_after_qc)

head(seurat_after_qc[[]])

# see if the dimensions of the RNA and SCT assays differ
dim(seurat_after_qc@assays$RNA) - dim(seurat_after_qc@assays$SCT)
```

Is there a change? 
> Answer: yes, new Assay called SCT 
> Note: the counts in the SCT assay are the corrected counts, not the raw counts 

Are there new columns in the meta.data?
> Answer: yes, we can see that the SCT function has recalculated library size based on corrected depth 

Why are the dimensions different between the SCT and RNA assays?
> Answer: there are fewer genes in the SCT assay vs RNA assay 
> This is because sctransform runs vst under the hood (see ?sctransform::vst)

# Exercise 2

## Visualisation

- The library size and number of features detected per cell is already present in the Seurat object.

- When you run `SCTransform` you get two new variables for library size and features for SCT normalisation

 - Use the function `VlnPlot()` to display compare RNA assay vs SCT assay library size in one plot 
 
 - Vlnplot of features detected in RNA vs SCT  in a single violin plot.

- Check to see how you can have fixed ylims in the `VlnPlot()` function

```{r}
VlnPlot(
  object = seurat_after_qc,
  features = c(
    "nCount_RNA", 
    "nCount_SCT"),
  pt.size = 0,
  same.y.lims = TRUE
)
```

Why do the 2 plots look different?
> Answer: We set same.y.lims = TRUE so that the y axis has the same scale for both plots, this allows us to see that the nCount_SCT data has a lot less variation than nCount_RNA 
> However, we can see that there is still variation within the nCount_SCT data (technical effects have been removed, but biological signals are still there)

## Bonus-
- Visualise the distribution of SCT corrected count data for a feature of your choice.
- Visualise the distribution of SCT logs counts in the data slot for a feature (e.g. LYZ)

- Visualise the distribution of SCT scale.data count feature of your choice. 

lets choose LYZ like day 1

```{r}
# set the default assay to SCT 
DefaultAssay(seurat_after_qc) <- "SCT"

# corrected counts are stored in the counts slot
ggplot_lyz_corrected <-
  ggplot(FetchData(seurat_after_qc, vars = "LYZ", slot = "counts"), aes(LYZ)) +
  geom_histogram(fill = "grey", color = "black", bins = 100) +
  coord_cartesian(ylim = c(0, 500)) +
  cowplot::theme_cowplot()

# normalised counts are stored in the data slot
ggplot_lyz_normalised <-
  ggplot(FetchData(seurat_after_qc, vars = "LYZ", slot = "data"), aes(LYZ)) +
  geom_histogram(fill = "grey", color = "black", bins = 100) +
  coord_cartesian(ylim = c(0, 500)) +
  cowplot::theme_cowplot()

# scaled counts are stored in the scale.data slot 
ggplot_lyz_scaled <-
  ggplot(FetchData(seurat_after_qc, vars = "LYZ", slot = "scale.data"), aes(LYZ)) +
  geom_histogram(fill = "grey", color = "black", bins = 100) +
  coord_cartesian(ylim = c(0, 500)) +
  cowplot::theme_cowplot()


cowplot::plot_grid(ggplot_lyz_corrected,
                   ggplot_lyz_normalised,ggplot_lyz_scaled,
                   ncol = 1)
```
Use the function `VariableFeatures` to pull out the 1:10 the variable genes after SCT and compare to 1:10 from the RNA assay

Do we need to change any arguments to get the variables genes specific to the SCT or RNA assay?

```{r}
VariableFeatures(seurat_after_qc, assay = "RNA") %>% 
  head(n=10)

VariableFeatures(seurat_after_qc, assay = "SCT") %>% 
  head(n=10)
```

## Bonus 

Check to see how similar/different the variable genes between the 2 normalisation methods are

```{r}

```

# Exercise 3


## Dimensionality reduction on SCT transformed data

- Run a principal component analysis and UMAP on the Seurat object.
- Check the Default assay

- Do we want to change the `reduction.name` argument so that we can still keep the RNA assay based PCA results?
  
```{r}
DefaultAssay(seurat_after_qc) <- "SCT"

# here we are setting the reduction name differently to the default ("pca") to prevent this PCA from overwriting the PCA from yesterday (which has the default name)

seurat_after_qc <- RunPCA(seurat_after_qc, reduction.name = "sct.pca" )
```

Check to see what reductions are now present in the object

```{r}
Reductions(seurat_after_qc)
```

- First, visualise the amount of variance explained the top principal components for SCTransformed data (number of your choice).
  How many principal components would you use for downstream analyses?
  
- Do we need to specify the correction reduction?

```{r}
ElbowPlot(seurat_after_qc, ndims = 100, reduction = "sct.pca")
```

Why do we view the ElbowPlot?
> Answer: We use the ElbowPlot to choose the number of principal components to keep. 20 seems like a good number. 

- Run the UMAP technique on the first 20 principal components stored in the Seurat object and visualise the result as a scatterplot.

How can we change the reduction name from default "umap" to "sct.umap"

How can we specify that we want to use pca run on the SCT Assay (sct.pca) in the previous step?



```{r}
seurat_after_qc <- RunUMAP(
  seurat_after_qc,
  dims = 1:20,
  reduction = "sct.pca",
  reduction.name = "sct.umap"
)
```

use DimPlot() to plot the umap, what happens if you try to specify different reductions with UMAPPlot()?

compare RNA based umap with sct.umap

```{r, fig.width=10}
p1 <- DimPlot(seurat_after_qc, reduction = "umap") + 
  ggtitle("RNA UMAP")

p2 <- DimPlot(seurat_after_qc, reduction = "sct.umap") +
  ggtitle("SCT UMAP")

p1 + p2
```


# Exercise 4

## Clustering on SCTransformed data

- Then, compute the graph of nearest neighbours using the function `FindNeighbors()`.
  Which principal components are used by default?
  Instead, specify the number of components that you have chosen.
  Have you chosen the right `reduction`  and `assay`? 
  
```{r}
seurat_after_qc <-  
  FindNeighbors(
    object = seurat_after_qc,
    reduction = "sct.pca", 
    dims = 1:20)
```


- Finally, compute cluster labels.
  What is the default setting for the `resolution` argument?
  Instead, set it to `0.5`.
  Do you expect more or fewer clusters following that change?
  What other parameters would you also try to experiment with?
  - should we specify the `graph.name` ?

```{r}
seurat_after_qc <- 
  FindClusters(
    object = seurat_after_qc,
    resolution = 0.5, 
    graph.name = "SCT_snn")
```


Check cluster assignment between SCT and RNA workflow

if you use the same resolution = o.5 and dims as RNA workflow do you get the same number of cluster or more?

are cells in the same cluster across both RNA and SCT
```{r}
table()

```

- Visualise the SCT cluster labels on the SCT transformed UMAP scatter plot and the RNA cluster labels on the RNA umap
  How would you describe the agreement between the UMAP layout and the clustering results?= for SCT vs RNA Assay results
  

```{r}
p1 <- 
  DimPlot(object = seurat_after_qc,
          reduction = "umap",
          group.by = "RNA_snn_res.0.5") + 
  ggtitle("RNA UMAP")

p2 <- 
  DimPlot(object = seurat_after_qc,
          reduction = "sct.umap",
          group.by = "SCT_snn_res.0.5") + 
  ggtitle("SCT UMAP")

p1 + p2

```


Plot some known cell-type markers for PBMC datasets, does the SCT better seperate the celltypes?

CD14+ Monocyte : LYZ, CD14
CD16 Monocytes : FCGR3A, MS4A7

CD4 T :  CD4, IL76

CD8 T : CD8A, CD3D

NK : GNLY, GZMB,NKG7

B Cell :  MS4A1 , CD79A

DC : CST3, FCER1A

Platelets : PPBP


```{r, fig.width=10, fig.height=8}
FeaturePlot(seurat_after_qc, 
            features = c("LYZ", "FCGR3A", "CD4", "CD8A", "GNLY", "MS4A1", "CST3", "PPBP"), 
            reduction = "sct.umap")

FeaturePlot(seurat_after_qc, 
            features = c("LYZ", "FCGR3A", "CD4", "CD8A", "GNLY", "MS4A1", "CST3", "PPBP"), 
            reduction = "umap")
```

This shows us that with SCT we are better able to pick out the CD8 subtype. 

# Bonus exercise to try in your own time:: Pathway analysis on Cluster markers for all clusters


- Choose either RNA workflow based generated or SCT generated based  seurat marker results

- we will be using Gprofiler `gost()` function fot multiple gene lists at the same time

- First we need to  filter seurat get top significant (p_adj < 0.05)  upregulated genes with a LogFC threshold (decided by you) for each cluster , use `dplyr:: group_by()` and `dplyr::filter` to get gene list for each cluster
then only select cluster and gene

- We then use ` split()`  on the filtered_df to divide gene markers into list  of multiple character vectors containing genes split by cluster

- can refer to the pahtway analysis code from week 3 but for h.sapiens

```{r}
seurat_clusters_results_filtered <- seurat_markers_all_sct %>% dplyr::group_by() %>% filter() %>%select(cluster,gene)
seurat_clusters_list <- split(seurat_clusters_results_filtered,cluster)
```


 We then run pathway analysis using `gost()` with multi_query = TRUE
 
 - to generate all_genes_id , we use all genes present in either the RNA assay or SCT assay , as we have already filtered out lowly expressed genes not present in certain cells .
 
```{r}
# Choose Default assay based on if running pathway analyses on RNA or SCT results
DefaultAssay(seurat_after_qc) <- ""
# create a vector of  of all genes 
all_genes_id <- rownames(seurat_after_qc)

multi_gostquery_results_obj <- gost(query = seurat_clusters_list,
                                    organism = "hsapiens", 
                                    exclude_iea = TRUE, # change to True as have alot of results otherwise
                                    correction_method = "g_SCS",  # this is thier own multiple testing coorrection - actually more strignent then bonferroni or BH 
                                    custom_bg = all_gene_ids, # make sure this is all gene ids 
                                    multi_query = TRUE)
```


can you plot the results for different clusters together ?

```{r}

gostplot(multi_gostquery_results_obj, capped = TRUE, interactive = TRUE)
```



# Afternoon Session

Demultiplexing with hashtag oligos (HTOs)

Dataset : 12-HTO dataset from four human cell lines

* Dataset represent Data represent single cells collected from four cell lines: HEK, K562, KG1 and THP1
* Each cell line was further split into three samples (12 samples in total)
* Each sample was labeled with a hashing antibody mixture (CD29 and CD45), pooled, and run on a single lane of 10X.
* Based on this design, we should be able to detect doublets both across and within cell types


Load in the UMI matrix for the RNA data
```{r}
hto12.umis <- readRDS("/project/obds/shared/resources/4_r_single_cell/seurat_day_2/data/hto12_umi_mtx.rds")

# how many cells and genes do we have in the RNA matrix?

dim(hto12.umis)

## columns are cells, rows are genes 

```


load in the HTO natrix

```{r}
hto12.htos <- readRDS("/project/obds/shared/resources/4_r_single_cell/seurat_day_2/data/hto12_hto_mtx.rds")

# what is the structure of the hto12.htos object? (what is the rows and what is the columns?)

dim(hto12.htos)

head(hto12.htos)

## Note that the columns and rows are the opoosite way round for the htos and umis 

# check the class of the hto object

class(hto12.htos) ## class is data.frame 



```


Now we only want to subset to those cell barcodes or  cells (actually called as cells by cellRanger or EmptyDrops on the gene expression data) which are detected by both RNA and HTO matrices? 

```{r}
joint.bcs <- intersect(rownames(hto12.htos), colnames(hto12.umis))

# check the class  of the joint.bcs object and how many cell barcodes do we have in common

class(joint.bcs) # class is a character vector 

length(joint.bcs) # we have 25088 barcodes common between the htos and umis (the length of the list)

```

# subset  the RNA matrix  to only the `joint.bcs` cell barcodes

```{r}
hto12.umis.common <- hto12.umis[,joint.bcs]

# check the dims
dim(hto12.umis.common)

## we have lost a lot of the cell barcodes
```




- Create  a Seurat object with the RNA (UMI count matrix) containing the `joint_bcs`

 Name the object `hto12_object`
 
 Include features detected in at least 3 cells,
 
 and cells where at least 200 features detected
 
- add HTO data as another assay to `hto12_object` 




```{r}
hto12_object <- CreateSeuratObject(
  counts = hto12.umis.common,
  project = "hto_umis",
  min.cells = 3,
  min.features = 200)

# Normalise with log normalization ,find variable genes and Scale RNA data

hto12_object <- NormalizeData(hto12_object)

hto12_object <- FindVariableFeatures(hto12_object)

hto12_object <- ScaleData(hto12_object)

```



we only want to subset the hto matrix to those cell barcodes which are now in the `hto12_object` Seurat object and  make sure that the features only consist of the HTO tags

Now look at the function `CreateAssayObject()` to  add the subsetted  HTO  matrix to the already created  `hto12_object` seurat object as a new assay called `HTO`

What is the expected format of the input matrix for `CreateAssayObject()`?

> Answer: features x cells (opposite way )

is our subsetted `hto12.htos.common` in the right format? if not what do we do to get it in the right format before adding it as another assay?

> Answer : we need to transpose the columns and the rows 


```{r}
# we want to subset the hto12.htos object such that  the rows or cell barcodes are only the same ones as the hto12_object and colnames are only consist of the 12 HTO tags
# we then want to transpoose this matrix, so now our cell barcodes are colnames and HTO tags are rownames 

# subset HTO matrix to only the barcodes in the hto12_object and keep only the first 12 columns as these contain the HTO tags 
# add t() to transpose the columns and rows (ensure the right way round for the CreateAssayObject function)

hto12.htos.common <-
  t(hto12.htos[colnames(hto12_object), 1:12])


```



Do we want to do any further filtering on the HTO object?

Answer : no - never filter out cells based on the HTO object 

Now add a new assay called "HTO" to the `hto12_object` seurat object

```{r}

hto12_object[["HTO"]] <- CreateAssayObject(counts = hto12.htos.common)

#check the dim of the seurat object
 dim(hto12_object)
```


## Normalise the HTO data 

Here we will use the CLR transformation with margin =1 (Default setting)

```{r}
# check the Default Assay
DefaultAssay(hto12_object) <- "HTO"

hto12_object <- NormalizeData(hto12_object, assay = "HTO", normalization.method = "CLR" , margin=1)
```


## Demultiplex cells based on HTO enrichment

Here we use Seurat Function `HTODemux()` to assign single cells to their original samples

```{r}
hto12_object <- HTODemux(hto12_object)
```

checkout the metadata column of the `hto12_object` , try to read the `HTODemux()` results output summary in the `Value` section to understand the results

```{r}
head(hto12_object[[]])

# hash.ID tells us which sample (HTO) that the cell belongs to 
# HTO_classification.global tells us whether the cells are doublets, singlets, or negative 
```



## Visualise the Demultiplexing results

we can visualise how many cells are classified as singlets, doublets and negative/ambiguous cells

check the meta.data, which column do we want for this information?

```{r}
table(hto12_object$HTO_classification.global)


```

## Visualize enrichment for selected HTOs with ridge plots

Plot the max HTO signal  for one of the HTO of each of the 4 cell lines (HEK, K562, KG1 and THP1) features with ridge plots using the `RidgePlot()` function

Plot Max HTO signal

```{r, fig.height=8, fig.width=10}
# Change the identities of the seurat object to the relevant metadata column

Idents(hto12_object) <- "hash.ID"
  
# use a ridgeline plot to see how different HTOs are assigned and whether there is overlap between cells assigned to different samples 
RidgePlot(
  object = hto12_object,
  features = c("HEK-A", "THP1-A", "K562-A", "KG1-A"),
  ncol=2)

```



## Visualize pairs of HTO signals to confirm mutual exclusivity in singlets  between the same celline

a) plot scatter plot of 2 HTOs within the same cell line e.g. HEK, colour by single/doublet/negative status

b) plot scatter plot of 2 HTOs within the same cell line e.g. HEK, colour by HTO_maxID 

c) plot scatter plot of 2 HTOs within the same cell line e.g. HEK, colour by HTO_secondID

use the function `FeatureScatter()`

```{r}
DefaultAssay(hto12_object) <- "HTO"

# a) 
FeatureScatter(
  object = hto12_object,
  feature1 = "HEK-A",
  feature2 = "HEK-B",
  group.by = "HTO_classification.global"
)

# b) 
FeatureScatter(
  object = hto12_object,
  feature1 = "HEK-A",
  feature2 = "HEK-B",
  group.by = "HTO_maxID"
)
```

What do you notice ?

1) 
2)



## Bonus Exercise

Plot scatter plot of 2 HTOs across different cell lines e.g. K562 vs KG1 and colour by (single/doublet/negative status) and HTO_max ID
 
```{r}

```
 
 
 
 
 
 
## Compare number of RNA UMIs for singlets, doublets and negative cells
 
What is a suitable plot for such comparisons?
 
> Answer: 

```{r}
Idents(hto12_object) <- "HTO_classification.global"

```

Question: what do you notice?

> Answer: 


## Visualize HTO signals in a heatmap , lookup `HTOHeatmap()`

```{r, fig.width=8}

HTOHeatmap(hto12_object)
```

What do you notice?

> Answer: we can see that the classification algorithm did a good job - most cells are assigned to a specific HTO (cell type) and there are not too many negatives. A reasonable number of cells have been identified as doublets. 


Generate a two dimensional tSNE or umap embedding for HTOs. Here we are grouping cells by singlets and doublets ONLY for simplicity.

Do we need to subset our object? If so what are we subsetting out?

> Answer:



```{r}

Idents(hto12_object) <- 
hto12_object.subset <- subset()
```


## Run UMAP/TSNE
 
What assay are we running UMAP/tsne for ?

look up the arguments in `RunUMAP()` and/or `RunTSNE()` functions

check which arguments in RunUMAP/RunUMAP/RunTSNE can be used to change the name of the reduction from defauult name of pca/umap/tsne to custom name

before we Run UMAP, we need to scale and run PCA like we did in the normal single cell workflow


> Answer:


```{r}
# Calculate a tSNE & UMAP embedding of the HTO data
DefaultAssay(hto12_object.subset) <- "HTO"




hto12_object.subset <- RunUMAP()

```

Check the Reductions in the object

```{r}
Reductions()
```


Plot the UMAP/tsne for the HTO assay
* which reduction shall we plot?

• colour by if singlet/doublet

• colour by HTO final classification results (hash.ID)

* check the arguments on how label the clusters by the cluster identity

* can chnage the label size?


```{r}

```

What do you notice about the cluustering on tthe UMAP/tsne, does the number of clusters mean anything?

> Answer: 

 
What do you notice about the cloud of cells surrounding each cluster?

> Answer: 


## Bonus exercise

### You can also visualize the more detailed classification result by group.by 
HTO_maxID before plotting. 

```{r}

```

What happens if you group.by the UMAP/TSNE plot by HTO..maxID?

> Answer: 


## Bonus Exercise (if we have time)


Cluster and visualize cells using the usual scRNA-seq workflow, and examine for the potential presence of batch effects.

Do we need to rerun `FindVariableFeatures()` and `ScaleData()` again?

> Answer: 

What other steps  do we need run to get viuslise our RNA data as umap/t-SNE? coloured by doublets/singlets/ and celltypes

> Answer: 


```{r}
DefaultAssay(hto12_object.subset) <- "RNA"
# Run PCA on most variable features
hto12_object.subset <- 
hto12_object.subset <- 


hto12_object.subset <- RunPCA(hto12_object.subset)
hto12_object.subset <- RunUMAP(hto12_object.subset, dims = 1:8)

```



# Plot RNA based UMAP

group.by hash.ID 

create a new seurat object  meta.data column called _cell_line , which removes "_A or _B or _C " in the hash.ID and replaces it with "", to create a new meta.data with only the cell-line info

```{r}
#we create another metadata column based on the hash.id column, where we gsub the HTO tag info (-A,-B,-C) for each cell line to plot only the cell lien names to see if we have batch effect

hto12_object.subset$cell_line <- gsub(pattern = "[-ABC]")
```


```{r}
DimPlot()
```

What does our RNA based clustering on the UMAP/T-SNE show?

> Answer: 
        


## Bonus exercise (try in your own time)

create a second seurat object based , using the code above, and rerun the `HTODemux()` with a different value of pasitive quantile.

try to check if the classificatiion changes massively if you adjusted the threshold for classification by playing around with the `positive.quantile` argument from the default.


