---
title: "single_cell_seurat_1"
author: "Hannah Pook"
date: '2022-05-23'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(Seurat)
library(tidyverse)
```

# Exercise 1

## Import scRNA-seq data and create a Seurat object

- Load the `Seurat` package.

```{r}
library(Seurat)
```

- Use the function `Read10X()` to import data in the directory `filtered_feature_bc_matrix/` 
  as an object named `read10x_data`.
  What class of object does the function return?

```{r}
read10x_data <- Read10X("/project/obds/shared/resources/4_r_single_cell/pbmc_5k/filtered_feature_bc_matrix/")
```

```{r}
class(read10x_data)
```

> Answer:
> The function returns a large dgCMatrix

- Have a look at the object and its structure (e.g., first 15 rows and 6 columns).
  What is a sparse matrix and how does it store data?

```{r}

```

```{r}

```

> Answer:
> 

- How many features and barcodes (i.e., cells) are present in the data set?

```{r}
dim(read10x_data)
```

> Answer:
> 

- Create a Seurat object using the function `CreateSeuratObject()`
  and the object `read10x_data`.
  Name the object `seurat_object`.
  Include features detected in at least 3 cells,
  and cells where at least 200 features detected.
  Name the project `pbmc5k`.
  How many features and barcodes are left in the Seurat object?

```{r}
seurat_object <- CreateSeuratObject(
  counts = read10x_data,
  project = "5k cells",
  assay = "RNA", 
  min.cells=3, 
  min.features=200)

seurat_object
```

> Answer:
> A smaller number of features are seen here 

- How many features and cells were filtered out when you created the Seurat object?

```{r}
dim(seurat_object) - dim(read10x_data)
```

> We discarded 14501 features and 55 cells 

# Exercise 2

## Accessing the contents of a Seurat object

- Query the name of the default assay in the Seurat object.

```{r}
DefaultAssay(seurat_object)
```

- List the names of assays available in the Seurat object.

```{r}
Assays(seurat_object)
```

- Display the first six rows and six columns of the RNA assay data.
  What function do you use?
  Which arguments do you need to supply, and which ones are optional?

```{r}
GetAssayData(seurat_object, slot="counts")[1:6,1:6]
```

> Answer:
> 

- Display the entire data.frame of per-cell metadata (first six rows).
  What column names do you see?

```{r}
head(seurat_object[[]])
```

> Answer:
> orig.ident gives the project name that we gave (as we did not give metadata)
> nCount_RNA gives the number of counts for each cell
> nFeature_RNA gives the number of features with non-zero counts 

- Fetch one column of metadata using `[[`.
  What type of object do you get back?

```{r}
FetchData(seurat_object, vars="nCount_RNA")

# another way of doing the same thing 
head(seurat_object[["nCount_RNA"]])
```

> Answer:
> Using the square bracket we get a data.frame back - useful for ggplot 

- Instead,fetch the same column of metadata using `$`.
  What type of object do you get back this time?

```{r}
head(seurat_object$nCount_RNA)
```

> Answer:
> This time we get back a vector 

- Use the function `FetchData()` to access the library size and expression of the feature named `"LYZ"` (first six rows).
  What type of object do you get back?

```{r}
head(FetchData(seurat_object, vars = c("nCount_RNA", "LYZ")))
```

> 

# Demo

## Common operations on Seurat objects

`WhichCells()` returns the names of cells that match a logical expression.

```{r}
WhichCells(seurat_object, expression = LYZ > 500)
```

`VariableFeatures()` returns the names of variable features (for a given assay, if computed).

```{r}
VariableFeatures(seurat_object)
```

`subset()` returns a new Seurat object restricted to certain features and cells.

```{r}
subset(
    x = seurat_object,
    cells = WhichCells(seurat_object, expression = LYZ > 500),
    features = VariableFeatures(object = seurat_object)
)
```

# Exercise 3

## Quality control and visualisation

- The library size and number of features detected per cell is already present in the Seurat object.
  Use the function `VlnPlot()` to display them in a single violin plot.

```{r}
VlnPlot(seurat_object, features=c("nCount_RNA", "nFeature_RNA"), pt.size=0)

#pt.size=0 to get rid of the points 
```

- Use the function `PercentageFeatureSet()` to compute the fraction of reads
  assigned to mitochondrial genes in each cell.
  Store the metric in the cell metadata of the Seurat object, under the name `"percent_mt"`.
  Visualise this new metric alongside the previous two in a new violin plot.

```{r}
percent_mt <- PercentageFeatureSet(seurat_object, pattern="MT-")

seurat_object$percent_mt <- percent_mt

VlnPlot(seurat_object, features=c("nCount_RNA", "nFeature_RNA", "percent_mt"), pt.size=0)
```

- Visualise a scatter plot of the proportion of mitochondrial UMIs against the library size in each cell.

```{r}
FeatureScatter(
  object=seurat_object,
  feature1="percent_mt", 
  feature2="nCount_RNA"
)
```

- Create a new Seurat object, called `seurat_after_qc`, that is subsetted to cells that have more than 4,500 UMI counts, less than 15% of UMI counts assigned to mitochondrial features, and more than 1,500 features detected.
  How many cells were removed in this step?

```{r}
seurat_after_qc <- subset(seurat_object, 
                          subset = nCount_RNA > 4500 & 
                            percent_mt < 15 &
                            nFeature_RNA > 1500)

seurat_after_qc
```

```{r}
dim(seurat_after_qc) - dim(seurat_object)
```

> Answer:
> We kept all features, but discarded 896 cells 

# Exercise 4

## Normalisation

- Normalise the RNA assay of the Seurat object (after quality control) using the `"LogNormalize"` method.

```{r}
seurat_after_qc <- NormalizeData(
  seurat_after_qc,
  normalization.method = "LogNormalize")
```

## Bonus

- Visualise the distribution of raw counts and normalised data for a feature of your choice.

```{r}
ggplot_lyz_raw <-
  ggplot(FetchData(seurat_after_qc, vars = "LYZ", slot = "counts"), aes(LYZ)) +
  geom_histogram(fill = "grey", color = "black", bins = 100) +
  coord_cartesian(ylim = c(0, 500)) +
  cowplot::theme_cowplot()

ggplot_lyz_normalised <-
  ggplot(FetchData(seurat_after_qc, vars = "LYZ", slot = "data"), aes(LYZ)) +
  geom_histogram(fill = "grey", color = "black", bins = 100) +
  coord_cartesian(ylim = c(0, 500)) +
  cowplot::theme_cowplot()

cowplot::plot_grid(ggplot_lyz_raw, ggplot_lyz_normalised, ncol = 1)
```

# Exercise 5

## Variable features and scaling

- Identify variable features in the normalised RNA assay of the Seurat object.
  Use the `"vst"` method and select the 2,000 most variable features.
  What does this subsetting do, and what are our motivations for doing it?

```{r}
seurat_after_qc <- FindVariableFeatures(
  seurat_after_qc,
  selection.method="vst",
  nfeatures=2000
)
```

> Answer:
> Using this function we identify the 2000 most variable features (genes)
> We need to select enough variable features in order to be able to identify different cell clusters downstream (even for rare cell types), but don't want to include too many as matrices will become very large and be difficult to store/work with 

- What is the function to display the name of variable features in a Seurat object (e.g., first 10)?
  How can you control which assay the variable features are pull from?

```{r}
head(VariableFeatures(seurat_after_qc), n=10)
```

> Answer:
> VariableFeatures displays the variable features 

- Use the function `VariableFeaturePlot()` to visualise the scatter plot of standardised variance against average expression.
  How would you use this plot?

```{r}
VariableFeaturePlot(seurat_after_qc)
```

> Answer:
> We can use the plot to identify how many genes seem to have highly variable expression and might be interesting 

- Scale the normalised RNA assay of the Seurat object, regressing the library size and the fraction of UMI counts assigned to mitochondrial features.
  What are the motivations for removing those two sources of variation?

```{r}
seurat_after_qc <- 
  ScaleData(seurat_after_qc,
            vars.to.regress = c("nCount_RNA", "percent_mt"),
            features=VariableFeatures(seurat_after_qc))
```

> Answer:
> 

# Exercise 6

## Dimensionality reduction

- Run a principal component analysis on the Seurat object.
  Which features are used by the method in the default settings?
  How could you change this?
  How do you read the message output of the function `RunPCA()`?

```{r}
seurat_after_qc <- RunPCA(seurat_after_qc)
```

> Answer:
> For each principle component it is printing the 30 genes which are most associated with it 

- List the names of dimensionality reduction results available in the Seurat object.

```{r}
Reductions(seurat_after_qc)
```

- Use `PCAPlot()` or `DimPlot()` to produce a scatterplot of the first and second PCA components.

```{r}
PCAPlot(seurat_after_qc)
```

## Bonus

- Make a scatterplot of the first and second PCA components yourself using `ggplot2`.

```{r}
# Use this code chunk to prepare a data.frame for ggplot2
pca_data <- FetchData(   )
head(pca_data)
```

```{r}
ggplot(pca_data,   ) +
    geom_point(size = 0.2) +
    cowplot::theme_cowplot()
```

- Visualise the amount of variance explained the top principal components (number of your choice).
  How many principal components would you use for downstream analyses?

```{r}
ElbowPlot(seurat_after_qc, ndims=50)
```

> 

- Run the UMAP technique on your selected number of principal components and visualise the result as a scatterplot.

```{r}
seurat_after_qc <- RunUMAP(seurat_after_qc, dims=1:20)

UMAPPlot(seurat_after_qc)
```

# Exercise 6

## Clustering

- Compute the graph of nearest neighbours using the function `FindNeighbors()`.
  Which principal components are used by default?
  Instead, specify the number of principal components that you have chosen earlier (20)

```{r}
seurat_after_qc <- FindNeighbors(seurat_after_qc, dims=1:20)
```

> Answer: 10 principal components are used by default, but we want to use 20 (the same number we chose from the elbow plot earlier)
> 
> The help page states that the function `FindNeighbors()` uses principal components 1 through 10, by default.

- What are the names of the nearest neighbour graphs that are now stored in the Seurat object?

```{r}
Graphs(seurat_after_qc)
```

- Finally, compute cluster labels.
  What is the default setting for the `resolution` argument?
  Instead, set it to `0.5`.
  Do you expect more or fewer clusters following that change?
  What other parameters would you also try to experiment with?

```{r}
seurat_after_qc <- FindClusters(seurat_after_qc, resolution = 0.5)
```

> 

- Visualise the cluster labels on the UMAP scatter plot.
  How would you describe the agreement between the UMAP layout and the clustering results?

```{r}
UMAPPlot(seurat_after_qc, group.by = "RNA_snn_res.0.5")
```

> The clustering is generally good, but might have slightly over-clustered 

```{r}
saveRDS(seurat_after_qc, file = "seurat_after_qc.RDS")
```


# Exercise 7

## Identify cluster markers

Load object from yesterday

```{r}
seurat_after_qc <- readRDS("seurat_after_qc.RDS")
```

- Use the function `FindAllMarkers()` to identify
  positive markers for all clusters,
  filtering markers that are detected in at least 25% of the cluster,
  and with a log fold-change greater than `0.25`.
  Assign the result to an object named `seurat_markers_all`.
  What is the class of that object?
  How do you control the set of clusters that are used?

```{r}
seurat_markers_all <- FindAllMarkers(
  object = seurat_after_qc,
  logfc.threshold = 0.25,
  min.pct = 0.25,
  only.pos = TRUE
)

class(seurat_markers_all)
```

> Answer:
> Default is to use current identities, can set the identities before looking for markers 

- How do you read the contents of the object `seurat_markers_all`?
  How do you know which features are the markers of each cluster?

```{r}
head(seurat_markers_all)
```

> Answer:
> 

- Filter and display the top 10 markers for cluster 3.

```{r}
head(filter(seurat_markers_all, cluster==3), n=10)

# an alternative method 
seurat_markers_all %>% filter(cluster==3) %>% slice_max(order_by = avg_log2FC,n=10)
```

- Visualise the expression of the top 4 marker for cluster 3 on a UMAP layout.

```{r}
FeaturePlot(
  object = seurat_after_qc,
  features = c("GZMK", "CCL5", "CST7", "GZMA")
)
```

- Visualise the expression of those same 4 marker as a violin plot.
  Do you have any particular preference between the two types of plots?

```{r}
VlnPlot(object = seurat_after_qc,
  features = c("GZMK", "CCL5", "CST7", "GZMA"))
```

> Answer:
> 

- Use `DoHeatmap()` to visualise the top 10 (positive) markers for each cluster.
  Hint: you may want to use the function `dplyr::group_by()`.

```{r}
markers_top10_clusters <- seurat_markers_all %>% 
  group_by(cluster) %>% 
  slice_max(order_by = avg_log2FC,n=10)
  
DoHeatmap(
  object = seurat_after_qc,
  features = markers_top10_clusters$gene)

```

```{r}
# an alternative method, selecting by lowest p value (i.e. most significant) instead of highest fold change 
markers_top10_clusters_new <- seurat_markers_all %>% 
  group_by(cluster) %>% 
  slice_min(order_by = p_val_adj, n=10, with_ties=FALSE)
  
DoHeatmap(
  object = seurat_after_qc,
  features = markers_top10_clusters_new$gene)
```